mag2exp/fmr.py

```diff
@@ -0,0 +1,116 @@
+"""FMR submodule.
+
+Module for calculation of Ferromagnetic Resonance related
+quantities.
+"""
+
+from typing import Optional, Tuple
+
+import discretisedfield as df
+import hvplot.xarray  # noqa
+import micromagneticdata as micd
+import numpy as np
+import scipy.fft as fft
+import xarray as xr
+
+
+def fmr(
+    drive: micd.Drive, init_field: Optional[df.Field] = None
+) -> Tuple[xr.DataArray, xr.DataArray]:
+    r"""
+    Computes the Ferromagnetic Resonance (FMR) power and phase spectra.
+
+    This function extracts the orientation of the magnetisation from the
+    provided drive object, optionally subtracts an initial field orientation
+    (if provided), and applies a Fourier transform to obtain the power
+    and phase spectra.
+
+    The Discrete Fourier Transform is computed at each spatial grid point for the
+    :math:`k`-th component of the magnetisation.
+    The corresponding power spectrum is given by:
+
+    .. math::
+
+        S_k(r_j, f) = \frac{1}{N} \left| F_k(r_j, f) \right|^2
+
+    where :math:`F_k(r_j, f)` is the Fourier transform of the magnetisation component
+    :math:`M_k(r_j, t)`.
+
+    The phase spectrum, which provides spatially resolved phase information,
+    is given by:
+
+    .. math::
+
+        \phi_k(r_j, f) = \arg F_k(r_j, f).
+
+    where :math:`\arg` is the phase angle of the complex Fourier coefficient.
+
+    For further details on numerical FMR analysis in micromagnetics see,
+    A. Baker et al., J. Magn. Magn. Mater. 421, 428 (2017)
+
+    Parameters
+    ----------
+    drive : micromagneticdata.Drive
+        A micromagnetic drive signal object containing time-dependent
+        magnetisation data.
+    init_field : discretisedfield.Field, optional
+        A reference field whose orientation is subtracted before processing.
+
+    Returns
+    -------
+    power : xr.DataArray
+        The computed power spectrum (squared magnitude of the FFT).
+    phase : xr.DataArray
+        The computed phase spectrum (FFT phase angle).
+    """
+    if not isinstance(drive, micd.Drive):
+        raise TypeError(
+            "The 'drive' parameter must be an instance of micromagneticdata.Drive."
```

Comment:
I would add (find this more convenient when running into the problem):
```suggestion
            f"The 'drive' parameter must be an instance of micromagneticdata.Drive, not {type(drive)=}"
```